### ISSUE TITLE
Take into account Julia's compilation time

### ISSUE THREAD

**Issue by balenamiaa** [hanabi1224/Another-Benchmarks-Game#46]
I am pretty sure you're not taking measures for Julia's compilation time or startup time. The memory usage and slow performance for Julia are absurd. It doesn't make sense to treat it like Python. For example, using Julia's [BenchmarkTools.jl](https://github.com/JuliaCI/BenchmarkTools.jl.git) -- `push!(ARGS, "4000"); @benchmark(include("pydigits.jl"))` inside a julia session -- instead of hyperfine "julia -O3 pydigits.jl 4000" results in a difference of 379.271 ms, 1.15 GiB and 1.333 s ±  0.008 s.

**Comment by hanabi1224** [comment]
Thanks for the feedback!  Currently, the framework is designed to be language-agnostic, it prefers standalone executables if AOT / bundling is possible. I'm not expert with Julia, it will be awesome if you can point me to some docs on doing AOT compilation of julia to eliminate compilation time, thanks!

**Comment by balenamiaa** [comment]
I took at the Python source code for pidigits, and Julia's. You've chosen the best contender on the original benchmarksgame for Python meanwhile there are more optimized versions like [this](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/pidigits-julia-3.html) for Julia.

As for precompilation, there's [PackageCompiler.jl](https://github.com/JuliaLang/PackageCompiler.jl), which will "AoT" compile. I've tried it locally and it leads to an improvement of about ~2.23x for pidigits.

> Take java as example, java allows ppl to precompile source code into bytecode (which requires JDK), then run bytecode on JVM with JRE only.

This doesn't apply to Julia in any way. The JIT compilation in Julia is very different from other languages, especially Java and C#. For one Julia does not have a virtual machine that interprets bytecode. Julia has a lot more in common with languages like C++ than managed ones for this particular aspect, in that it directly compiles to optimized machine code. The fact that it is based on LLVM should've made this clear. Also, the base library is already precompiled into a sysimage that ships with Julia, so a "hello world" won't take that much memory 😛 . In fact this is what PackageCompiler.jl does. I personally do not program microservices but I am sure a lot have in the Julia community for HPC clusters, and there seems to be no problem. Even though you're probably aware, but compilation only has to happen once at function boundaries when it is given a new type(basically specialized machine code is generated whenever you call a generic function on a new type that it hasn't been ran through before); this can happen at any stage of your code, and usually adds a few hundred milliseconds of delay the first time. For any long running process, this shouldn't be a problem at all and if it is, there's PackageCompiler.jl for now.

**Comment by hanabi1224** [comment]
@balenamiaa Awesome, the relocatable app is exactly what I'm looking for, will try to switch to it when I get a chance. Looking forward to its perf gain

**Comment by hanabi1224** [comment]
@balenamiaa  BTW, regarding the fast version of julia pidigits, it's using gmp :)  You may notice that none of the versions I take use native bindings, the whole point of pidigits is to measure perf of big int implementation of a language, any solutions that binds to gmp are almost benchmarks of gmp itself, which does not make any sense to me.

**Comment by hanabi1224** [comment]
Also after taking some time looking into julia, I actually got some questions on your previous answer.

> The JIT compilation in Julia is very different from other languages, especially Java and C#. For one Julia does not have a virtual machine that interprets bytecode. Julia has a lot more in common with languages like C++ than managed ones for this particular aspect, in that it directly compiles to optimized machine code. The fact that it is based on LLVM should've made this clear.

Do you mean that Julia does JIT complication for code that run only once? 

From what I understand as long as julia uses JIT, it's nowhere close to C/C++/rust/go... which are fully AOT compiled.  Please note that JVM / dotnet CLR uses JIT compliation as well for hot code path, it's actually more advanced to automatically determine running code in interpreter mode or jit mode to achive best online performance than having only either.

BTW, regarding VM, julia doc says it JIT compiles julia IR with LLVM in runtime, which is actually more close to java/dotnet mode.  The only difference is that julia does not precompile code to its IR, while java/dotnet does (.class,.dll are IL/IR to their VM/runtime).  And VM is kind of a naming flavor, the part of julia compiler that JIT compiles IR to machine code has no fundamental difference  from JVM or dotnet CLR. The reason why term JVM is used everywhere is just that JVM is an abstract spec that has many different concrete implementations, if one day julia has more alternative JIT compilers maybe ppl will call it VM as well.

Regarding  PackageCompiler.jl, I learnt that julia's stdlib is already precompiled to sys.so and shipped with sdk. And the fact that I could not find an easy one line bash command that compiles single .jl source code file into sysimage and run systemimage directly, makes me feel that it's a non-goal for julia-lang to directly AOT compile source code into executables (stdlib is already AOT compiled tho), so I decide not to do that for now. (But PRs are always welcome.)

**Comment by balenamiaa** [comment]
> Please note that JVM / dotnet CLR uses JIT compliation as well for hot code path, it's actually more advanced to automatically determine running code in interpreter mode or jit mode to achive best online performance than having only either.

Well, you just answered your question. Julia does not interpret code and to say that it's merely the same as java, except you can't "precompile" to bytecode is simply wrong.

> BTW, regarding VM, julia doc says it JIT compiles julia IR with LLVM in runtime, which is actually more close to java/dotnet mode.

Most programming languages out there at some point compile down to IR, that doesn't make it similar to java/dotnet.

> VM is kind of a naming flavor, the part of julia compiler that JIT compiles IR to machine code has no fundamental difference from JVM or dotnet CLR. The reason why term JVM is used everywhere is just that JVM is an abstract spec that has many different concrete implementations, if one day julia has more alternative JIT compilers maybe ppl will call it VM as well.

This is just simply wrong. A Virtual Machine(VM) isn't just a naming flavor... there's a reason nobody is mentioning a VM for Julia, because there is none. 

> The only difference is that julia does not allow user to precompile code to its IR and run IR directly, while java/dotnet does (.class,.dll are IL/IR to their VM/runtime).

This truly suggests a very grave misunderstanding of Julia... now why would Julia "allow" that?

The way .NET and JVM languages work(or used to work more accurately) is quite similar to python. The difference is that python interprets its source code directly, meanwhile the former ones first "compile" their source code down to an intermediate representation that is low-level, quite similar to assembly. Then they provide a virtual machine that runs on the target computer that interprets that IR and runs it. JIT was added as an optimization later on and the way it works is that it compiles whatever piece of IR that runs a lot, or so-called "hotspots", to machine code and instead of interpreting, it hands over to that machine code for that part.

Julia is pretty much similar to AOT compiled languages in that it directly runs in machine code, there's no interpreting or some low-level bytecode(Julia IR is just simply lowered code, as most AOT compiled languages have one). The difference is that the machine code is actually compiled on the run(not runtime) instead of being compiled once and bundled into an executable or some object file. 

"Since Julia uses JIT, Julia generates native machine code directly, before a function is first run (i.e. a different approach than compiling to bytecode, that you distribute by default, to be run on a virtual machine (VM), as with e.g. Java/JVM; then translated from the bytecode while running, as done by Dalvik on older versions of Android)."

"Julia uses a just-in-time (JIT) compiler that is referred to as "just-ahead-of-time" (JAOT) in the Julia community, as Julia compiles all code (by default) to machine code before running it."

These are a few quotes from Wikipedia.

BTW, it seems you keep mentioning docker over and over. This isn't a benchmark related to docker at all, it isn't a be-all-and-end-all.

**Comment by balenamiaa** [comment]
As for hello world's memory usage, proper benchmarking shows "56.600 μs (8 allocations: 176 bytes)". True, that isn't a language-agnostic way of benchmarking, but it is the proper way especially in Julia's case. What you're most likely getting the high amount of memory from is the startup of the Julia REPL, which is like a sophisticated application, so you aren't truly benchmarking "hello world". This is why the Julia memory usage for every other benchmark in the list never drops below ~150 mb; because that's what it requires to startup the Julia REPL.

**Comment by hanabi1224** [comment]
> What you're most likely getting the high amount of memory from is the startup of the Julia REPL, which is like a sophisticated application, so you aren't truly benchmarking "hello world".

I may need to clarify that, the purpose of hello world is to measure startup time, mem consumption of a most simple application in a language, It makes no sense to measure a single print function.

Please note that for an industry programming language, it's critical to optimize startup time / JIT time. e.g. ruby has --jit-opt--<> to optimize when JIT should kick in, dotnet has PublishReadyToRun flag to optimize startup time.

> Julia is pretty much similar to AOT compiled languages in that it directly runs in machine code, there's no interpreting or some low-level bytecode(Julia IR is just simply lowered code, as most AOT compiled languages have one). The difference is that the machine code is actually compiled on the run(not runtime) instead of being compiled once and bundled into an executable or some object file.

Well, this really sounds like double standards to me. AOT is a choice that sacrifices compilation time to achieve best startup speed, interpreter or JIT are choices to achieve quick compilation with sacrificing some runtime performance. When a language claims to be AOT, it should provide an obvious way to produce standalone executables, which Julia does not, then to be fair, its JIT compilation time cost should definitely be counted just as what we do to all other non-AOT compilers. 

> "Since Julia uses JIT, Julia generates native machine code directly, before a function is first run (i.e. a different approach than compiling to bytecode, that you distribute by default, to be run on a virtual machine (VM), as with e.g. Java/JVM; then translated from the bytecode while running, as done by Dalvik on older versions of Android)."

This is just a description of how JIT works in Julia, which I don't see any differences from JIT in other languages, and I cannot get how this statement makes Julia more similar to AOT (like c) than other JITs (like java). Think about why it's call JUST-IN-TIME, it means it happens in runtime, versus AHEAD-OF-TIME.  One cannot claim its JUST-IN-TIME complication should be taken out of measurement without providing AHEAD-OF-TIME choice : )

> "Julia uses a just-in-time (JIT) compiler that is referred to as "just-ahead-of-time" (JAOT) in the Julia community, as Julia compiles all code (by default) to machine code before running it."

JAOT sounds like a term to re-phrase the less optimal JIT compiler I described above, which does not make it an excuse to not count its Just-In-Time compilation.  A scenario that makes JAOT valid I could think of is, use JIT for local development to save compilation time, then provide full compilation to achieve the best production performance, otherwise, the term does not make any sense to me.

Again, the benchmark here is designed to be language-agnostic and should apply the same standard to all languages, and the goal of each problem is to reflect the true fact of some aspect.  Take hello world, again, as example, it tells ppl that if they'd like to build a simple micro service with low mem consumption and time-to-start, which languages are good choices which are not,.
